File: python/median_filter.py
import cv2
import numpy as np
import time
import argparse


# Function to apply median filter to a single channel with black edges
def apply_median_filter(input_channel, kernel_size):
    edge = kernel_size // 2
    # Initialize output with zeros (black edges)
    output = np.zeros_like(input_channel)
    # start_time = time.time()
    for x in range(edge, input_channel.shape[1] - edge):
        for y in range(edge, input_channel.shape[0] - edge):
            neighbors = input_channel[y - edge:y + edge + 1, x - edge:x + edge + 1]
            output[y, x] = np.median(neighbors)
    # end_time = time.time()
    # print(f"loop time: {end_time - start_time} seconds")
    return output

def main():
    parser = argparse.ArgumentParser(description='Apply median filter to an image.')
    parser.add_argument('image_file', help='Path to the image file')

    args = parser.parse_args()

    # Read the image using the provided file path
    image_file_name = args.image_file
    image = cv2.imread(image_file_name, cv2.IMREAD_COLOR)

    # # Read the image
    # image = cv2.imread('../resources/noise_intro_2.jpg', cv2.IMREAD_COLOR)

    # Split the image into its color channels
    channels = cv2.split(image)

    start_time = time.time()

    # Apply median filter to each channel
    filtered_channels = [apply_median_filter(ch, 5) for ch in channels]  # Kernel size is 5

    end_time = time.time()
    print(f"Filtering time: {end_time - start_time} seconds")
    
    # Merge the channels back
    filtered_image = cv2.merge(filtered_channels)

    # Save the images before and after filtering
    cv2.imwrite('before_py.jpg', image)
    cv2.imwrite('after_py.jpg', filtered_image)

if __name__ == "__main__":
    main()

File: c/median_filter.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;
using namespace std::chrono;

// Function to apply median filter to a single channel
void applyMedianFilter(const cv::Mat& input, cv::Mat& output, int kernelSize) {
    int edge = kernelSize / 2;
    std::cout << "Inside Medail Fitler\n";
    for (int y = edge; y < input.rows - edge; y++) {
        for (int x = edge; x < input.cols - edge; x++) {
            // std::cout << x << " " << y << "\n";
            std::vector<uchar> neighbors;
            for (int dy = -edge; dy <= edge; dy++) {
                for (int dx = -edge; dx <= edge; dx++) {
                    neighbors.push_back(input.at<uchar>(y + dy, x + dx));
                }
            }
            std::nth_element(neighbors.begin(), neighbors.begin() + neighbors.size() / 2, neighbors.end());
            output.at<uchar>(y, x) = neighbors[neighbors.size() / 2];
        }
    }
}

int main(int argc, char *argv[]) {

    string image_name;
    if(argc > 1)
        image_name = argv[1];
    else
        image_name = "sp_img_gray_noise_heavy.png";
    cv::Mat image = cv::imread("resources/" + image_name, cv::IMREAD_COLOR);


    cv::Mat frame, filteredFrame;
    cv::Mat channels[3], outputChannels[3];
    int N = 1; // Replace with the number of frames you want to process

    // Save the frame before filtering
    std::string filenameBefore = "before.jpg" ;
    cv::imwrite(filenameBefore, image);

    // Split the frame into its color channels
    cv::split(image, channels);

    for (int i = 0; i < 3; i++) {
        outputChannels[i] = cv::Mat(image.rows, image.cols, CV_8UC1);
    }

    // Apply median filter to each channel    
    auto start_mf = high_resolution_clock::now();
    for (int i = 0; i < 3; i++) {
        applyMedianFilter(channels[i], outputChannels[i], 5); // Kernel size is 5
    }
    auto end_mf = high_resolution_clock::now();

    // Merge the channels back
    cv::merge(outputChannels, 3, filteredFrame);


    // Save the frame after filtering
    std::string filenameAfter = "after.jpg";
    cv::imwrite(filenameAfter, filteredFrame);
    
    auto filter_duration = duration_cast<microseconds>(end_mf - start_mf).count();
    std::cout << "filter time   : " << filter_duration << " us" << std::endl;


    return 0;
}